homeassistant/components/mqtt/notify.py

```diff
@@ -0,0 +1,397 @@
+        vol.Optional(CONF_TARGETS, default=[]): cv.ensure_list,
+        vol.Optional(CONF_TITLE, default=notify.ATTR_TITLE_DEFAULT): cv.string,
+        vol.Optional(CONF_RETAIN, default=mqtt.DEFAULT_RETAIN): cv.boolean,
+    }
+)
+
+DISCOVERY_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Optional(CONF_DEVICE): MQTT_ENTITY_DEVICE_INFO_SCHEMA,
+    },
+    extra=vol.REMOVE_EXTRA,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class MqttNotificationConfig(TypedDict, total=False):
+    """Supply service parameters for MqttNotificationService."""
+
+    command_topic: str
+    command_template: Template
+    encoding: str
+    name: str | None
+    qos: int
+    retain: bool
+    targets: list
+    title: str
+    device: ConfigType
+
+
+async def async_initialize(hass: HomeAssistant) -> None:
+    """Initialize globals."""
+    await async_setup_reload_service(hass, DOMAIN, PLATFORMS)
+    hass.data.setdefault(MQTT_NOTIFY_SERVICES_SETUP, {})
+
+
+def _check_notify_service_name(hass: HomeAssistant, config: ConfigType) -> str | None:
+    """Check if the service already exists or else return the service name."""
+    service_name = slugify(config[CONF_NAME])
+    has_services = hass.services.has_service(notify.DOMAIN, service_name)
+    services = hass.data[MQTT_NOTIFY_SERVICES_SETUP]
+    if service_name in services.keys() or has_services:
+        _LOGGER.error(
+            "Notify service '%s' already exists, cannot register service",
+            service_name,
+        )
+        return None
+    return service_name
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up MQTT notify service dynamically through MQTT discovery."""
+    await async_initialize(hass)
+    setup = functools.partial(_async_setup_notify, hass, config_entry=config_entry)
+    await async_setup_entry_helper(hass, notify.DOMAIN, setup, DISCOVERY_SCHEMA)
+
+
+async def _async_setup_notify(
+    hass,
+    legacy_config: ConfigType,
+    config_entry: ConfigEntry,
+    discovery_data: dict[str, Any],
+):
+    """Set up the MQTT notify service with auto discovery."""
+    config = DISCOVERY_SCHEMA(discovery_data[ATTR_DISCOVERY_PAYLOAD])
+    discovery_hash = discovery_data[ATTR_DISCOVERY_HASH]
+
+    if not (service_name := _check_notify_service_name(hass, config)):
+        async_dispatcher_send(hass, MQTT_DISCOVERY_DONE.format(discovery_hash), None)
+        clear_discovery_hash(hass, discovery_hash)
+        return
+
+    device_id = None
+    if CONF_DEVICE in config:
+        await _update_device(hass, config_entry, config)
+        device_registry = await hass.helpers.device_registry.async_get_registry()
+        device = device_registry.async_get_device(
+            {(DOMAIN, id_) for id_ in config[CONF_DEVICE][CONF_IDENTIFIERS]},
+            {tuple(x) for x in config[CONF_DEVICE][CONF_CONNECTIONS]},
+        )
+        device_id = device.id
```

Comment:
```suggestion
        device_id = _update_device(hass, config_entry, config).id
```

Reply:
This code part is also in the tag platform, if this works, we could cleanup there as well.